File: Algorithms/p006_ZigZag_Conversion/p006_ZigZag_Conversion.py
class Solution(object):
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        if numRows<=1:
            return s
        n = len(s)
        m = numRows*2-2
        l = [s[0:n:m]]
        for i in range(1,numRows-1):
            j = i
            d = m-i*2
            while j<n:
                l.append(s[j])
                j += d
                d = m-d
        l.append(s[numRows-1:n:m])
        return ''.join(l)

File: Algorithms/p006_ZigZag_Conversion/p006_ZigZag_Conversion.cpp
class Solution {
public:
    string convert(string s, int numRows) {
        if(numRows<=1) return s;
        int n = s.size();
        int m = numRows*2-2;
        string res;
        for(int i=0;i<n;i+=m){
            res += s[i];
        }
        for(int i=1;i<numRows-1;++i){
            for(int j=i,d=m-i*2;j<n;j+=d,d=m-d){
                res += s[j];
            }
        }
        for(int i=numRows-1;i<n;i+=m){
            res += s[i];
        }
        return res;
    }
};
